crates/rust-analyzer/src/main_loop.rs

```diff
@@ -436,6 +436,10 @@ impl GlobalState {
                 handlers::handle_matching_brace(s.snapshot(), p)
             })?
             .on_sync::<lsp_ext::MemoryUsage>(|s, p| handlers::handle_memory_usage(s, p))?
+            .on_sync::<lsp_types::request::Completion>(handlers::handle_completion)?
```

Comment:
> on_sync

I had to make both synchronous, because I cannot put `ImportEdit` into `GlobalStateSnapshot`: its `ImportScope` contains rowan syntax nodes that have `Rc` in them and it's not possible to pass it between threads safely.

Since current insert calculations may take up to ~200ms, I get an "overly long loop" pop-up on the majority of the calculation resolve requests, yet it's not notable in the UI much, since occurs during the completion selection only.

Reply:
I wonder if we can maybe use a `SyntaxNodePtr` here to get around the sync problem? Not sure if this is a good application for it

Reply:
I'm not quite sure how I'm able to retrieve the `root: &SyntaxNode` on `handle_completion_resolve` step to reverse `SyntaxNodePtr` into a `SyntaxNode`, do you have any ideas?

I can look into this later.

Reply:
Can you reconstruct it in a cheap manner if you know that you're handling the "add import" case?

Reply:
I've understood your idea about having zero extra state, not sure I can implement it simply: `ImportScope` needs `hir::Semantics` and an ast node (or a pointer) of something close to the original completion and `ModPath` that does not allow itself to be created from strings intentionally.

Reply:
Yet the resolve request is now async, no more irritating pop-up, yet the resolve method handler is the same.

Reply:
It's been a while since I've looked in that space but why can't you construct the `hir::Semantics` with the `db` and pass enough information through `data` to reconstruct (or find) the ast node?

Reply:
No state now.